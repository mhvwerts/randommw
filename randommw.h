/*==========================================================================
 * randommw.h 
 * combines Doornik's zigrandom.h and zignor.h, with modifications and
 * extensions
 *
 * Werts, 2024
 *==========================================================================*/




/* from zigrandom.h */
#ifndef ZIGRANDOM_H
#define ZIGRANDOM_H

#include <stdint.h>

#define M_RAN_INVM30	9.31322574615478515625e-010			  /* 1.0 / 2^30 */
#define M_RAN_INVM32	2.32830643653869628906e-010			  /* 1.0 / 2^32 */
#define M_RAN_INVM48	3.55271367880050092936e-015			  /* 1.0 / 2^48 */
#define M_RAN_INVM52	2.22044604925031308085e-016			  /* 1.0 / 2^52 */
#define M_RAN_INVM64	5.42101086242752217004e-020			  /* 1.0 / 2^64 */

#define RANDBL_32new(iRan1)                   \
    ((int)(iRan1) * M_RAN_INVM32 + (0.5 + M_RAN_INVM32 / 2))
#define RANDBL_48new(iRan1, iRan2)            \
    ((int)(iRan1) * M_RAN_INVM32 + (0.5 + M_RAN_INVM48 / 2) + \
        (int)((iRan2) & 0x0000FFFF) * M_RAN_INVM48)
#define RANDBL_52new(iRan1, iRan2)            \
    ((int)(iRan1) * M_RAN_INVM32 + (0.5 + M_RAN_INVM52 / 2) + \
        (int)((iRan2) & 0x000FFFFF) * M_RAN_INVM52)

/* xoshiro256+ */
void RanSetSeed_xoshiro256p(uint64_t uSeed);
uint32_t IRan_xoshiro256p(void);
double DRan_xoshiro256p(void);


/* MWC8222 George Marsaglia */

/* void 	GetInitialSeeds(uint32_t auiSeed[], int32_t cSeed,
	                    uint32_t uiSeed, uint32_t uiMin);
*/

void RanSetSeed_MWC8222(uint64_t uSeed);
uint32_t IRan_MWC8222(void);
double DRan_MWC8222(void);
double DRan_MWC_52(void);

/* plug-in RNG */
typedef double 		( * DRANFUN)(void);
typedef uint32_t 	( * IRANFUN)(void);
typedef void   		( * RANSETSEEDFUN)(uint64_t);

void    RanSetRan(const char *sRan);
void    RanSetRanExt(DRANFUN DRanFun, IRANFUN IRanFun, RANSETSEEDFUN RanSetSeedFun);
double  DRanU(void);
uint32_t  IRanU(void);
void    RanSetSeed(uint64_t uSeed);

/* normal probabilities */
double  DProbNormal(double x);

/* polar standard normal RNG */
double  DRanNormalPolar(void);
double  FRanQuanNormal(void);
double  DRanQuanNormal(void);

#endif /* ZIGRANDOM_H */


/* from zignor.h */
double  DRanNormalZig(void);
double  DRanQuanNormalZig(void);

/* xoshiro256 */

/* further extensions */
void  RanInit(uint64_t uSeed);
