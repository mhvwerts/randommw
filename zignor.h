void    RanNormalSetSeedZig(int *piSeed, int cSeed);
double  DRanNormalZig(void);

double  DRanQuanNormalZig(void);

