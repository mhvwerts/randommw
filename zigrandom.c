/*==========================================================================
 *  Modified version of zigrandom.c
 *  Martinus H. V. Werts, 2024
 *
 *  - Fixed a gcc warning
 *  - Added MWC_52 random number generator, which is a full-precision
 *    generator of random doubles in (0, 1) from pairs of 64-bit unsigned
 *    integers coming from MWC_8222 (= MWC_256)
 *==========================================================================*/

/*==========================================================================
 *  This code is Copyright (C) 2005, Jurgen A. Doornik.
 *  Permission to use this code for non-commercial purposes
 *  is hereby given, provided proper reference is made to:
 *		Doornik, J.A. (2005), "An Improved Ziggurat Method to Generate Normal
 *          Random Samples", mimeo, Nuffield College, University of Oxford,
 *			and www.doornik.com/research.
 *		or the published version when available.
 *	This reference is still required when using modified versions of the code.
 *  This notice should be maintained in modified versions of the code.
 *	No warranty is given regarding the correctness of this code.
 *==========================================================================*/

#include <limits.h>
#include <stdint.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

#include "zigrandom.h"

/*---------------------------- GetInitialSeeds -----------------------------*/
void GetInitialSeeds(uint32_t auiSeed[], int32_t cSeed,
	uint32_t uiSeed, uint32_t uiMin)
{
	int i;
	uint32_t s = uiSeed;									/* may be 0 */

	for (i = 0; i < cSeed; )
	{	/* see Knuth p.106, Table 1(16) and Numerical Recipes p.284 (ranqd1)*/
		s = 1664525 * s + 1013904223;
		if (s <= uiMin)
			continue;
        auiSeed[i] = s;
		++i;
    }
}
/*-------------------------- END GetInitialSeeds ---------------------------*/


/*------------------------ George Marsaglia MWC ----------------------------*/
#define MWC_R  256
#define MWC_A  809430660ull // unsigned long long is 64 bits
#define MWC_AI 809430660
#define MWC_C  362436
static uint32_t s_uiStateMWC = MWC_R - 1;
static uint32_t s_uiCarryMWC = MWC_C;
static uint32_t s_auiStateMWC[MWC_R];

void RanSetSeed_MWC8222(int *piSeed, int cSeed)
{
	s_uiStateMWC = MWC_R - 1;
	s_uiCarryMWC = MWC_C;
	
	if (cSeed == MWC_R)
	{
		int i;
		for (i = 0; i < MWC_R; ++i)
		{
			s_auiStateMWC[i] = (uint32_t)piSeed[i];
		}
	}
	else
	{
		GetInitialSeeds(s_auiStateMWC, MWC_R, piSeed && cSeed ? piSeed[0] : 0, 0);
	}
}
uint32_t IRan_MWC8222(void)
{
	uint64_t t;

	s_uiStateMWC = (s_uiStateMWC + 1) & (MWC_R - 1);
	t = MWC_A * s_auiStateMWC[s_uiStateMWC] + s_uiCarryMWC;
	s_uiCarryMWC = (uint32_t)(t >> 32);
	s_auiStateMWC[s_uiStateMWC] = (uint32_t)t;
    return (uint32_t)t;
}
double DRan_MWC8222(void)
{
	uint64_t t;

	s_uiStateMWC = (s_uiStateMWC + 1) & (MWC_R - 1);
	t = MWC_A * s_auiStateMWC[s_uiStateMWC] + s_uiCarryMWC;
	s_uiCarryMWC = (uint32_t)(t >> 32);
	s_auiStateMWC[s_uiStateMWC] = (uint32_t)t;
	return RANDBL_32new(t);
}
double DRan_MWC_52(void)
/* Generate random doubles with full-precision 52-bit mantissa using MWC8222 */
{
	uint64_t t1, t2;

	s_uiStateMWC = (s_uiStateMWC + 1) & (MWC_R - 1);
	t1 = MWC_A * s_auiStateMWC[s_uiStateMWC] + s_uiCarryMWC;
	s_uiCarryMWC = (uint32_t)(t1 >> 32);
	s_auiStateMWC[s_uiStateMWC] = (uint32_t)t1;
	
	s_uiStateMWC = (s_uiStateMWC + 1) & (MWC_R - 1);
	t2 = MWC_A * s_auiStateMWC[s_uiStateMWC] + s_uiCarryMWC;
	s_uiCarryMWC = (uint32_t)(t2 >> 32);
	s_auiStateMWC[s_uiStateMWC] = (uint32_t)t2;
	
	return RANDBL_52new(t1, t2);
}
/*----------------------- END George Marsaglia MWC -------------------------*/


/*------------------- normal random number generators ----------------------*/
static int s_cNormalInStore = 0;		     /* > 0 if a normal is in store */

/* Default MWC_52 uniform generator (MWC8222 with 52 bits mantissa) */
static DRANFUN s_fnDRanu = DRan_MWC_52;
static IRANFUN s_fnIRanu = IRan_MWC8222;
static RANSETSEEDFUN s_fnRanSetSeed = RanSetSeed_MWC8222;

double  DRanU(void)
{
    return (*s_fnDRanu)();
}
uint32_t IRanU(void)
{
    return (*s_fnIRanu)();
}
void    RanSetSeed(int *piSeed, int cSeed)
{
   	s_cNormalInStore = 0;
	(*s_fnRanSetSeed)(piSeed, cSeed);
}
void    RanSetRan(const char *sRan)
{
   	s_cNormalInStore = 0;
	
	/* BEGIN if ... else if ... else block */
	if (strcmp(sRan, "MWC8222") == 0)
	{
		s_fnDRanu = DRan_MWC8222;
		s_fnIRanu = IRan_MWC8222;
		s_fnRanSetSeed = RanSetSeed_MWC8222;
	}
	else if (strcmp(sRan, "MWC_52") == 0)
	{
		s_fnDRanu = DRan_MWC_52;
		s_fnIRanu = IRan_MWC8222;
		s_fnRanSetSeed = RanSetSeed_MWC8222;
	}
	else // DEFAULT
	{
		s_fnDRanu = NULL;
		s_fnIRanu = NULL;
		s_fnRanSetSeed = NULL;
	}
	/* END if ... else if ... else block */
}
static uint32_t IRanUfromDRanU(void)
{
    return (uint32_t)(UINT_MAX * (*s_fnDRanu)());
}
static double DRanUfromIRanU(void)
{
    return RANDBL_32new( (*s_fnIRanu)() );
}
void    RanSetRanExt(DRANFUN DRanFun, IRANFUN IRanFun, RANSETSEEDFUN RanSetSeedFun)
{
	s_fnDRanu = DRanFun ? DRanFun : DRanUfromIRanU;
	s_fnIRanu = IRanFun ? IRanFun : IRanUfromDRanU;
	s_fnRanSetSeed = RanSetSeedFun;
}
/*---------------- END uniform random number generators --------------------*/


/*----------------------------- Polar normal RNG ---------------------------*/
#define POLARBLOCK(u1, u2, d)	              \
	do                                        \
	{   u1 = (*s_fnDRanu)();  u1 = 2 * u1 - 1;\
		u2 = (*s_fnDRanu)();  u2 = 2 * u2 - 1;\
		d = u1 * u1 + u2 * u2;                \
	} while (d >= 1);                         \
	d = sqrt( (-2.0 / d) * log(d) );       	  \
	u1 *= d;  u2 *= d

static double s_dNormalInStore;

double  DRanNormalPolar(void)                         /* Polar Marsaglia */
{
    double d, u1;

    if (s_cNormalInStore)
        u1 = s_dNormalInStore, s_cNormalInStore = 0;
    else
    {
        POLARBLOCK(u1, s_dNormalInStore, d);
        s_cNormalInStore = 1;
    }

return u1;
}

#define FPOLARBLOCK(u1, u2, d)	              \
	do                                        \
	{   u1 = (float)((*s_fnDRanu)());  u1 = 2 * u1 - 1;\
		u2 = (float)((*s_fnDRanu)());  u2 = 2 * u2 - 1;\
		d = u1 * u1 + u2 * u2;                \
	} while (d >= 1);                         \
	d = sqrt( (-2.0 / d) * log(d) );       	  \
	u1 *= d;  u2 *= d

static float s_fNormalInStore;

double  FRanNormalPolar(void)                         /* Polar Marsaglia */
{
    float d, u1;

    if (s_cNormalInStore)
        u1 = s_fNormalInStore, s_cNormalInStore = 0;
    else
    {
        POLARBLOCK(u1, s_fNormalInStore, d);
        s_cNormalInStore = 1;
    }

return (double)u1;
}
/*--------------------------- END Polar normal RNG -------------------------*/

/*------------------------------ DRanQuanNormal -----------------------------*/
static double dProbN(double x, int fUpper)
{
    double p;  double y;  // int fnegative = 0; // mw231228 remove gcc warning

    if (x < 0)
        x = -x, fUpper = !fUpper; // fnegative = 1,  // mw231228 remove gcc warning
    else if (x == 0)
        return 0.5;

    if ( !(x <= 8 || (fUpper && x <= 37) ) )
        return (fUpper) ? 0 : 1;

    y = x * x / 2;

    if (x <= 1.28)
    {
        p = 0.5 - x * (0.398942280444 - 0.399903438504 * y /
            (y + 5.75885480458 - 29.8213557808 /
            (y + 2.62433121679 + 48.6959930692 /
            (y + 5.92885724438))));
    }
    else
    {
        p = 0.398942280385 * exp(-y) /
            (x - 3.8052e-8 + 1.00000615302 /
            (x + 3.98064794e-4 + 1.98615381364 /
            (x - 0.151679116635 + 5.29330324926 /
            (x + 4.8385912808 - 15.1508972451 /
            (x + 0.742380924027 + 30.789933034 /
            (x + 3.99019417011))))));
    }
    return (fUpper) ? p : 1 - p;
}
double  DProbNormal(double x)
{
    return dProbN(x, 0);
}
double  DRanQuanNormal(void)
{
	return DProbNormal(DRanNormalPolar());
}
double  FRanQuanNormal(void)
{
	return DProbNormal(FRanNormalPolar());
}
/*----------------------------- END DRanQuanNormal -------------------------*/
