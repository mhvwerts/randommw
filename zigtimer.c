/*==========================================================================
 *  This code is Copyright (C) 2005, Jurgen A. Doornik.
 *  Permission to use this code for non-commercial purposes
 *  is hereby given, provided proper reference is made to:
 *		Doornik, J.A. (2005), "An Improved Ziggurat Method to Generate Normal
 *          Random Samples", mimeo, Nuffield College, University of Oxford,
 *			and www.doornik.com/research.
 *		or the published version when available.
 *	This reference is still required when using modified versions of the code.
 *  This notice should be maintained in modified versions of the code.
 *	No warranty is given regarding the correctness of this code.
 *==========================================================================*/
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "zigtimer.h"

/*--------------------------- timer functions ------------------------------*/

#if defined(__linux__) | defined (__APPLE__)
#include <sys/times.h>
#include <time.h>

clock_t unix_clock (void)
{
    struct tms buf;
    if (times (&buf) < 0)
        return clock();

    return buf.tms_utime + buf.tms_stime;
}

#define clock unix_clock
#include <unistd.h>
#ifdef CLOCKS_PER_SEC
#undef CLOCKS_PER_SEC /* to stop warnings, always 1000000 under POSIX */
#endif
#define CLOCKS_PER_SEC (sysconf (_SC_CLK_TCK))
#endif

static char s_sTime[24];
static clock_t s_clockStart, s_clockStop;
static double s_dClockPerHundred = 0;


void    StartTimer(void)
{
	if (s_dClockPerHundred == 0)
		s_dClockPerHundred = 100.0 / CLOCKS_PER_SEC;
		
	s_clockStop = s_clockStart = clock();
}
void    StopTimer(void)
{
    s_clockStop = clock();
}

static double  ClockStartDblHsec(void)
{
    return s_clockStart * s_dClockPerHundred;
}
static double  ClockStopDblHsec(void)
{
    return s_clockStop * s_dClockPerHundred;
}

static char *  GetTimeSpan(double dTime2, double dTime1)
{
    int m, s, h, len, bzeros;  unsigned long tnew;
    double  duration;
       
    duration = fabs(dTime2 - dTime1);             /* difference in hundreds */
    tnew = (unsigned long)(duration / 100.0);      /* difference in seconds */
    
    h = (int)(duration - 100.0 * tnew);                         /* hundreds */
    s = tnew % 60;  tnew /= 60;                                  /* seconds */
    m = tnew % 60;  tnew /= 60;             /* minutes, now tnew is # hours */
    
    len = bzeros = 0;
    if (tnew)
    {   sprintf(s_sTime, "%2d:", (int)tnew);  len = (int)strlen(s_sTime);
        bzeros = 1;
    }
    if (m || tnew)
    {   sprintf(s_sTime + len, bzeros ? "%02d:" : "%2d:", m);  len += 3;
        bzeros = 1;
    }
    sprintf(s_sTime + len, bzeros ? "%02d" : "%2d", s); len += 2;
    if (h)
        sprintf(s_sTime + len, ".%02d", h);

return s_sTime;
}
char * GetLapsedTime(void)
{
    return GetTimeSpan(ClockStopDblHsec(), ClockStartDblHsec());
}
/*-------------------------- END timer functions ---------------------------*/
