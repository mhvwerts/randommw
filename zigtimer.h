void 	StartTimer(void);
void 	StopTimer(void);
char * GetLapsedTime(void);

