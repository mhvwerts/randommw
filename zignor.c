/*==========================================================================
 *  Modified version of zignor.c
 *  Martinus H. V. Werts, 2024
 *
 *  - VIZIGNOR and related optimizations have been removed
 *==========================================================================*/

/*==========================================================================
 *  This code is Copyright (C) 2005, Jurgen A. Doornik.
 *  Permission to use this code for non-commercial purposes
 *  is hereby given, provided proper reference is made to:
 *		Doornik, J.A. (2005), "An Improved Ziggurat Method to Generate Normal
 *          Random Samples", mimeo, Nuffield College, University of Oxford,
 *			and www.doornik.com/research.
 *		or the published version when available.
 *	This reference is still required when using modified versions of the code.
 *  This notice should be maintained in modified versions of the code.
 *	No warranty is given regarding the correctness of this code.
 *==========================================================================*/
#include <limits.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "zigrandom.h"
#include "zignor.h"


/*------------------------------ General Ziggurat --------------------------*/
static double DRanNormalTail(double dMin, int iNegative)
{
	double x, y;
	do
	{	x = log(DRanU()) / dMin;
		y = log(DRanU());
	} while (-2 * y < x * x);
	return iNegative ? x - dMin : dMin - x;
}

#define ZIGNOR_C 128			       /* number of blocks */
#define ZIGNOR_R 3.442619855899	/* start of the right tail */
				   /* (R * phi(R) + Pr(X>=R)) * sqrt(2\pi) */
#define ZIGNOR_V 9.91256303526217e-3

/* s_adZigX holds coordinates, such that each rectangle has*/
/* same area; s_adZigR holds s_adZigX[i + 1] / s_adZigX[i] */
static double s_adZigX[ZIGNOR_C + 1], s_adZigR[ZIGNOR_C];

static void zigNorInit(int iC, double dR, double dV)
{
	int i;	double f;
	
	f = exp(-0.5 * dR * dR);
	s_adZigX[0] = dV / f; /* [0] is bottom block: V / f(R) */
	s_adZigX[1] = dR;
	s_adZigX[iC] = 0;

	for (i = 2; i < iC; ++i)
	{
		s_adZigX[i] = sqrt(-2 * log(dV / s_adZigX[i - 1] + f));
		f = exp(-0.5 * s_adZigX[i] * s_adZigX[i]);
	}
	for (i = 0; i < iC; ++i)
		s_adZigR[i] = s_adZigX[i + 1] / s_adZigX[i];
}
double  DRanNormalZig(void)
{
	unsigned int i;
	double x, u, f0, f1;
	
	for (;;)
	{
		u = 2 * DRanU() - 1;
		i = IRanU() & 0x7F;
		/* first try the rectangular boxes */
		if (fabs(u) < s_adZigR[i])		 
			return u * s_adZigX[i];
		/* bottom box: sample from the tail */
		if (i == 0)						
			return DRanNormalTail(ZIGNOR_R, u < 0);
		/* is this a sample from the wedges? */
		x = u * s_adZigX[i];		   
		f0 = exp(-0.5 * (s_adZigX[i] * s_adZigX[i] - x * x) );
		f1 = exp(-0.5 * (s_adZigX[i+1] * s_adZigX[i+1] - x * x) );
      	if (f1 + DRanU() * (f0 - f1) < 1.0)
			return x;
	}
}

void  RanNormalSetSeedZig(int *piSeed, int cSeed)
{
	zigNorInit(ZIGNOR_C, ZIGNOR_R, ZIGNOR_V);
	RanSetSeed(piSeed, cSeed);
}

/*--------------------------- END General Ziggurat -------------------------*/

/*--------------------------- functions for testing ------------------------*/
double  DRanQuanNormalZig(void)
{
	return DProbNormal(DRanNormalZig());
}

/*------------------------- END functions for testing ----------------------*/
