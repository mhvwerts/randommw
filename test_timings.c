#include <stdio.h>
#include <stdlib.h>

#include "zigrandom.h"
#include "zigtimer.h"
#include "zignor.h"


void TimerHeader() 
{
	printf("------------------------------------------------------------------\n");
	printf("%-30s%8s %22s %10s\n", "Name", "time", "mean", "reps");

}
void Timer(char *sName, double (*DRan)(), void (*DRanSeed)(int *, int), unsigned int cM) 
{
	unsigned int i;
	double ran, mean;
	
	printf("%-29s", sName);

	(*DRanSeed)(NULL, 0);

	StartTimer();
	for (i = 0, mean = 0.0; i < cM; ++i)
	{
		ran = (*DRan)();
		mean += ran;
	}
	mean /= cM;
	StopTimer();

	printf(" %8s %#22.15g %10d\n", GetLapsedTime(), mean, cM);
}

int main(void) 
{
	unsigned int cm = 1000000000;

	TimerHeader();

	Timer("Warming up",		DRan_MWC8222, 			RanSetSeed_MWC8222, cm/10); 
	Timer("MWC8222",		DRan_MWC8222, 			RanSetSeed_MWC8222, cm);

	RanSetRan("MWC8222");
	Timer("ZIGNOR",			DRanNormalZig, 			RanNormalSetSeedZig, cm);

	return 0;
}
