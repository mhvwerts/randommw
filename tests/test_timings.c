#include <stdio.h>
#include <stdlib.h>

#include "randommw.h"
#include "zigtimer.h"


void TimerHeader() 
{
	printf("------------------------------------------------------------------------\n");
	printf("%-30s%8s %22s %10s\n", "Name", "time", "mean", "reps");
	printf("------------------------------------------------------------------------\n");
}

void TimerFooter()
{
	printf("------------------------------------------------------------------------\n");
}

void Timer(char *sName, double (*DRan)(), void (*DRanSeed)(uint64_t), unsigned int cM) 
{
	unsigned int i;
	double ran, mean;
	
	printf("%-29s", sName);

	(*DRanSeed)(12345);

	StartTimer();
	for (i = 0, mean = 0.0; i < cM; ++i)
	{
		ran = (*DRan)();
		mean += ran;
	}
	mean /= cM;
	StopTimer();

	printf(" %8s %#22.15g %10d\n", GetLapsedTime(), mean, cM);
}


int main(void) 
{
	unsigned int cm = 1000000000;

	TimerHeader();

	Timer("Warming up",		DRan_MWC256, 			RanSetSeed_MWC256, cm/10); 
	Timer("MWC256",			DRan_MWC256, 			RanSetSeed_MWC256, cm);
	Timer("Xoshiro256+",	DRan_xoshiro256p,		RanSetSeed_xoshiro256p, cm);

	RanSetRan("MWC256");
	Timer("ZIGNOR MWC256",			DRanNormalZig, 			RanInit, cm);
	
	RanSetRan("Xoshiro256+");
	Timer("ZIGNOR Xoshiro256+",		DRanNormalZig, 			RanInit, cm);
	
	RanSetRan("MWC256");
	Timer("ZIGNOR MWC256 (again)",	DRanNormalZig, 			RanInit, cm);

	
	/* Code snippet to see if the program indeed crashes gracefully if
	 * wrong string passed to RanSetRan                          */
	/*
	RanSetRan("CRASH!! SEGV...");
	Timer("CRASH!!",	DRanNormalZig, 			RanInit, cm);
	*/
	
	TimerFooter();
	
	return 0;
}
