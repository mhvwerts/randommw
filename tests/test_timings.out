------------------------------------------------------------------------
Name                              time                   mean       reps
------------------------------------------------------------------------
Warming up                      0.20 s      0.499966335061385  100000000
MWC256                          2.02 s      0.499999113678946 1000000000
Lehmer64                        1.93 s      0.499991240100718 1000000000
Xoshiro256+                     1.82 s      0.500000988475582 1000000000
MELG19937                       2.82 s      0.499992994773625 1000000000
ZIGNOR MWC256                   6.99 s   3.51278967376399e-05 1000000000
ZIGNOR Lehmer64                 6.90 s   9.76531810136464e-07 1000000000
ZIGNOR Xoshiro256+              6.96 s   3.51948053403410e-05 1000000000
ZIGNOR MELG19937                9.58 s  -4.49744996458742e-06 1000000000
ZIGNOR MWC256 (again)           7.04 s   3.51278967376399e-05 1000000000
------------------------------------------------------------------------
